cmd/juju/application/deployer/deployer.go

```diff
@@ -54,25 +75,202 @@ func NewDeployerFactory(dep DeployerDependencies) DeployerFactory {
+			supportedJujuSeries: workloadSeries,
+			force:               d.force,
+			conf:                modelCfg,
+			fromBundle:          false,
+		}
+
+		if len(supportedSeries) == 0 {
+			logger.Warningf("%s does not declare supported series in metadata.yml", ch.Meta().Name)
+		}
+
+		seriesName, err = seriesSelector.charmSeries()
+		if err = charmValidationError(ch.Meta().Name, errors.Trace(err)); err != nil {
+			return "", "", errors.Trace(err)
+		}
+	}
+	return seriesName, imageStream, nil
+}
+
 // GetDeployer returns the correct deployer to use based on the cfg provided.
 // A ModelConfigGetter needed to find the deployer.
 func (d *factory) GetDeployer(cfg DeployerConfig, getter ModelConfigGetter, resolver Resolver) (Deployer, error) {
 	d.setConfig(cfg)
-	maybeDeployers := []func() (Deployer, error){
-		d.maybeReadLocalBundle,
-		func() (Deployer, error) { return d.maybeReadLocalCharm(getter) },
-		d.maybePredeployedLocalCharm,
-		func() (Deployer, error) { return d.maybeReadRepositoryBundle(resolver) },
-		d.repositoryCharm, // This always returns a Deployer
-	}
-	for _, d := range maybeDeployers {
-		if deploy, err := d(); err != nil {
+
+	// Determine the type of deploy we have
+	var dk DeployerKind
+
+	userCharmURL, err := resolveCharmURL(d.charmOrBundle, d.defaultCharmSchema)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if isLocalSchema(d.charmOrBundle) {
+		// Check if the path is ambiguous
+		_, fileStatErr := d.fileSystem.Stat(d.charmOrBundle)
+		if fileStatErr == nil && !charm.IsValidLocalCharmOrBundlePath(d.charmOrBundle) {
+			return nil, errors.Errorf(""+
+				"The charm or bundle %q is ambiguous.\n"+
+				"To deploy a local charm or bundle, run `juju deploy ./%[1]s`.\n"+
+				"To deploy a charm or bundle from CharmHub, run `juju deploy ch:%[1]s`.",
+				d.charmOrBundle,
+			)
+		}
+		// Check Local Bundle
+		if ds, localBundleDataErr := charm.LocalBundleDataSource(d.charmOrBundle); localBundleDataErr == nil {
+			// Set the deployer kind to localBundleDeployerType
+			dk = &localBundleDeployerType{DataSource: ds}
+		} else if !errors.Is(localBundleDataErr, errors.NotFound) {
+			// Only raise if it's not a NotFound.
+			// Otherwise, no need to raise, it's not a bundle,
+			// continue with trying for local charm.
+			return nil, errors.Annotatef(localBundleDataErr, "cannot deploy %v", d.charmOrBundle)
+		} else {
+			// Check Local Charm
+			charmOrBundle := d.charmOrBundle[6:]
+			seriesName, imageStream, seriesErr := d.determineSeriesForLocalCharm(charmOrBundle, getter)
+			if seriesErr != nil {
+				return nil, errors.Trace(seriesErr)
+			}
+			// Charm may have been supplied via a path reference.
+			ch, curl, err := corecharm.NewCharmAtPathForceSeries(charmOrBundle, seriesName, d.force)
+			// We check for several types of known error which indicate
+			// that the supplied reference was indeed a path but there was
+			// an issue reading the charm located there.
+			if corecharm.IsMissingSeriesError(err) {
```

Comment:
switches can be easier to read than a very long if/else if/self

Reply:
Not sure if that if block is suited for a switch. It's not something I wrote, but I might be able to change things around to refactor that a little bit 👍